packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -432,21 +432,42 @@ export default function makeKernelKeeper(storage) {
     storage.set(`${kernelSlot}.data.slots`, capdata.slots.join(','));
   }
 
-  function findPromisesDecidedByVat(vatID) {
-    const prefixKey = `${vatID}.c.p`;
-    const endKey = `${vatID}.c.q`;
-    const result = [];
-    for (const k of storage.getKeys(prefixKey, endKey)) {
+  function cleanupAfterTerminatedVat(vatID) {
+    insistVatID(vatID);
+    ephemeral.vatKeepers.delete(vatID);
+    const koPrefix = `${vatID}.c.o+`;
+    const kpPrefix = `${vatID}.c.p`;
+    const kernelPromisesToReject = [];
+    for (const k of storage.getKeys(`${vatID}.`, `${vatID}/`)) {
       // The store semantics ensure this iteration is lexicographic.  Any
       // changes to the creation of the list of promises need to preserve this
       // in order to preserve determinism.
-      const kpid = storage.get(k);
-      const p = getKernelPromise(kpid);
-      if (p.state === 'unresolved' && p.decider === vatID) {
-        result.push(kpid);
+      if (k.startsWith(koPrefix)) {
+        const koid = storage.get(k);
+        const ownerKey = `${koid}.owner`;
+        const ownerVat = storage.get(ownerKey);
+        if (ownerVat === vatID) {
+          storage.set(ownerKey, 'none');
+        }
+      } else if (k.startsWith(kpPrefix)) {
+        const kpid = storage.get(k);
+        const p = getKernelPromise(kpid);
+        if (p.state === 'unresolved' && p.decider === vatID) {
+          kernelPromisesToReject.push(kpid);
+        }
       }
+      storage.delete(k);
     }
-    return result;
+    // TODO: deleting entries from the dynamic vat IDs list requires a linear
+    // scan; this collection ought to be represented in a way that makes it
+    // efficient to remove an entry from it, though this should be OK as long as
+    // we keep the list short.
+    const DYNAMIC_IDS_KEY = 'vat.dynamicIDs';
+    const oldDynamicVatIDs = JSON.parse(getRequired(DYNAMIC_IDS_KEY));
+    const newDynamicVatIDs = oldDynamicVatIDs.filter(v => v !== vatID);
+    storage.set(DYNAMIC_IDS_KEY, JSON.stringify(newDynamicVatIDs));
```

Comment:
we should probably sort this list before writing it back, it needs to be deterministic

the same "lexicographic is easier to do in JS, but numeric would be nicer for a spec document and in future non-JS languages" issue is present for this sort. Use lexicographic for now, but we need to revisit before we lock down the consensus spec.

Reply:
In this case sorting should not be needed -- essentially all we're doing is removing an element from an array, which will preserve the existing order of the elements that were already there.

Reply:
Also, prior to deleting any entries from this list, the order is not lexicographic anyway.  The ordering is the temporal order that the vats were created.  Given that the computation which creates the vats is deterministic, the ordering of the list must be as well.

The lexicographic ordering thing had to do with multiple entries in the kernel DB being sequenced lexicographically by key, but this list is always stored as a single database entry with a single key, `vat.dynamicIDs`.